C++: Reshuffle IR directories

dave-bartolomeo opened the pull request:
This change rearranges the IR directory structure to be consistent between the three current IR flavors. It also makes "IR.qll" resolve to the most precise flavor of the IR, and makes that the only publicly exposed flavor.

There are no code changes here other than fixing up module references. The three commits can be reviewed one-by-one if that makes more sense.

I'm getting these changes in now before feature freeze so that master won't diverge too wildly as I do more IR work.

jbj commented on config/identical-files.json:
```diff
@@ -1,58 +1,58 @@
-        "cpp/ql/src/semmle/code/cpp/ssa/internal/aliased_ssa/Instruction.qll"
+        "cpp/ql/src/semmle/code/cpp/ir/internal/raw_ir/Instruction.qll",
+        "cpp/ql/src/semmle/code/cpp/ir/internal/ssa_ir/Instruction.qll",
+        "cpp/ql/src/semmle/code/cpp/ir/internal/aliased_ssa_ir/Instruction.qll"
```
Isn't the `_ir` suffix on the directories redundant? They're already below a directory named `ir`.

jbj commented on cpp/ql/src/semmle/code/cpp/ir/IRSanity.ql:
```diff
@@ -5,4 +5,4 @@
  */
 
-import internal.IRSanityImpl
+import internal.aliased_ssa_ir.IRSanity
```
Isn't this file and most other files in `ir` internal and should be moved to the `internal` dir? Some should probably be moved into the subdir of `internal` that's appropriate for the IR dialect they apply to. If users should only ever import `semmle.code.ir.IR`, then I'd say all other files are internal.

ian-semmle replied:
There's no need for users to import e.g. `semmle.code.cpp.Function` rather than `cpp` either, is there?

I would say that `.internal.` should be for modules that are only privately imported, and that users should not import. i.e. we can freely change them without breaking user code.

jbj replied:
> I would say that `.internal.` should be for modules that are only privately imported, and that users should not import. i.e. we can freely change them without breaking user code.

I can see we've used "internal" with that meaning everywhere except in the pyrameterized modules of Java/C++ data flow and C++ IR. I agree it's a useful distinction and that we should probably reserve "internal" to mean what it's always meant.

I still see value in discouraging users from importing things like `semmle.code.cpp.ir.MemoryAccessKind` directly, though. I think of that file name as internal (subject to change between releases) even if the file contents is exposed somewhere in the namespace. Maybe a directory named "implementation" would be appropriate for such code?

dave-bartolomeo replied:
`semmle/code/cpp/ir` now contains only three modules: `IR.qll`, `PrintIR.ql`, and `IRSanity.ql`, all of which forward to the corresponding `implementation/aliased_ssa` versions. Everything else is now in either `implementation` or `internal`.

rdmarsh2 commented:
This is currently targeted at master; should it go onto rc/1.18 instead?

dave-bartolomeo commented:
Retargeted to rc/1.18
I've made @jbj's suggested change to use "implementation" for public but not directly importable modules, and "internal" for modules that should not be exposed outside our library.

dave-bartolomeo commented:
My most recent push should fix the test failures. @jbj this is ready to merge as soon as the test run comes back green.

dave-bartolomeo commented:
Rebased, and added a commit to fix the deprecated predicate warnings that #141 was trying to fix. That commit conflicted with the reshuffling, and it only updated on copy of `PrintIRImpl.qll`.
